source/tests/system/nidaqmx_driver_api_tests.cpp

```diff
@@ -1439,10 +1477,109 @@ TEST_F(NiDAQmxDriverApiTests, ChannelWithDoneEventRegisteredTwice_RunCompleteFin
+  ASSERT_EQ(Completion(START_CALL_TAG, true), get_next_completion(cq));
+  reader->ReadInitialMetadata(READ_INITIAL_METADATA_TAG);
+  ASSERT_EQ(Completion(READ_INITIAL_METADATA_TAG, true), get_next_completion(cq));
+  cfg_samp_clk_timing(
+      create_cfg_samp_clk_timing_request(1000.0, Edge1::EDGE1_RISING, AcquisitionType::ACQUISITION_TYPE_CONT_SAMPS, 0UL));
+
+  start_task();
+  std::vector<RegisterEveryNSamplesEventResponse> responses;
+  RegisterEveryNSamplesEventResponse response;
+  for (auto i = 0UL; i < N_READS; ++i) {
+    reader->Read(&response, READ_TAG);
+    ASSERT_EQ(Completion(READ_TAG, true), get_next_completion(cq));
+    responses.push_back(response);
+    read_analog_f64(N_SAMPLES, N_SAMPLES);
+  }
+
+  EXPECT_THAT(responses, SizeIs(N_READS));
+  EXPECT_THAT(responses, Each(Property(&RegisterEveryNSamplesEventResponse::n_samples, Eq(N_SAMPLES))));
+  EXPECT_THAT(responses, Each(Property(&RegisterEveryNSamplesEventResponse::status, Eq(DAQMX_SUCCESS))));
+}
+
+// This test uses the async API in order to check whether the event stream is closed without blocking.
+TEST_F(NiDAQmxDriverApiTests, AsyncEveryNSamplesEventRegisteredWithWrongEventType_WaitForInitialMetadataAndReadStream_ReadCompletesImmediatelyWithError)
+{
+  const auto N_SAMPLES = 10UL;
+  const auto START_CALL_TAG = (void*)1;
+  const auto READ_INITIAL_METADATA_TAG = (void*)2;
+  const auto READ_TAG = (void*)3;
+  const auto FINISH_TAG = (void*)4;
+  create_ai_voltage_chan(0.0, 1.0);
+  ::grpc::ClientContext reader_context;
+  ::grpc::CompletionQueue cq;
+  auto request = create_register_every_n_samples_event_request(N_SAMPLES, EveryNSamplesEventType::EVERY_N_SAMPLES_EVENT_TYPE_TRANSFERRED_FROM_BUFFER);
+  std::unique_ptr<::grpc::ClientAsyncReader<RegisterEveryNSamplesEventResponse>> reader(
+    stub()->AsyncRegisterEveryNSamplesEvent(&reader_context, request, &cq, START_CALL_TAG));
+  ASSERT_EQ(Completion(START_CALL_TAG, true), get_next_completion(cq));
+
+  reader->ReadInitialMetadata(READ_INITIAL_METADATA_TAG);
+  ASSERT_EQ(Completion(READ_INITIAL_METADATA_TAG, true), get_next_completion(cq));
+  RegisterEveryNSamplesEventResponse response;
+  reader->Read(&response, READ_TAG);
+  Completion read_completion;
+  bool got_completion_without_blocking = try_get_next_completion(cq, read_completion, /* timeout */ std::chrono::milliseconds(0));
+
+  EXPECT_TRUE(got_completion_without_blocking);
+  EXPECT_EQ(Completion(READ_TAG, false), read_completion);  // read_completion should be NOT OK
+  ::grpc::Status status;
+  reader->Finish(&status, FINISH_TAG);
+  ASSERT_EQ(Completion(FINISH_TAG, true), get_next_completion(cq));
+  EXPECT_THROW({
+    try {
+      client::raise_if_error(status, reader_context);
+    }
+    catch (const client::grpc_driver_error& ex) {
+      expect_driver_error(ex, EVERY_N_SAMPS_TRANSFERRED_FROM_BUFFER_EVENT_NOT_SUPPORTED_BY_DEVICE_ERROR);
+      throw;
+    }
+  }, client::grpc_driver_error);
```

Comment:
FYI, I want to try cleaning up these `EXPECT_THROW` blocks by extracting an `EXPECT_THROW_GRPC_DRIVER_ERROR` macro. I'll post a separate PR.

Reply:
https://github.com/ni/grpc-device/pull/933